### PR TITLE
Week 4 -  add Sentiment Analysis to the comments 

### PR DESCRIPTION
**Pull request by ralucatudor**
Added color to the comment based on how negative or positive its text is - using Cloud Natural Language, _"which lets you analyze text and extract information like sentiment"_.

Added `sentimentScore` field to Comment class. Updated DataServlet.java to calculate comment `sentimentScore` in `doPost()` and also send it in `doGet()` along with the previous fields. 

Then set comments' color based on the `sentimentScore` which is a float between -1 and 1 (so the negative text is **red** - value **-1**, positive is **green** - value **1**, and in-between is **yellow** - value **0**). I blended the colors (see the interpolate functions) to get the exact color corresponding to the score - i.e. for a comment with the score 0.5 I get a color between yellow and green.

[Live server](http://rtudor-step-2020.appspot.com/)

### CODE REVIEW

File: portfolio/src/main/webapp/script.js [hunk 1]
```diff
@@ -170,6 +170,13 @@ function createCommentElement(comment) {
   const commentText = document.createElement('div');
   commentText.innerText = comment.text;
 
+  // Set the commentText's color property corresponding to the sentimentScore
+  const commentTextColor = getSentimentColor(comment.sentimentScore);
+  commentText.style.color = 'rgb(' + commentTextColor.r + 
+                            ',' + commentTextColor.g + 
+                            ',' + commentTextColor.b +
+                            ')';
```
Comment (skopecki) [inline portfolio/src/main/webapp/script.js]: You could also write
`` `rgb(${ commentTextColor.r },${ commentTextColor.g },${ commentTextColor.b })` ``

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -192,6 +199,56 @@ function createCommentElement(comment) {
...
 
+/**
+ * Returns a color from red to green as an RGB value, corresponding to the sentimentScore.
+ * @param {Number} sentimentScore - float between [-1.0, 1.0]
+ */
+function getSentimentColor(sentimentScore) {
+  // Initialize the colors
+  const red = {r : 255, g : 0, b : 0};
+  const yellow = {r : 210, g : 210, b : 0};
+  const green = {r : 0, g : 255, b : 0};
```
Comment (skopecki) [inline portfolio/src/main/webapp/script.js]: The pure red and green values are very aggressive and often not that easy to read. You could for example pick values from https://clrs.cc/

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -192,6 +199,56 @@ function createCommentElement(comment) {
...
+
+  // Handle wrong input cases
+  if (sentimentScore < -1) {
+    sentimentScore = -1;
+  }
+  if (sentimentScore > 1) {
+    sentimentScore = 1;
+  }
```
Comment (skopecki) [inline portfolio/src/main/webapp/script.js]: It's common to use Math.min and Math.max here:
` Math.min(1, Math.max(-1, sentimentScore))`

**Review by skopecki — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -108,4 +113,25 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+
+  /**
+   * Returns a value between -1 and 1, representing how negative or positive text is
+   */
+  private static float getSentimentScore(String text) throws IOException {
+    // Create a new Document that contains the parameter - text - as content
+    Document doc =
+        Document.newBuilder().setContent(text).setType(Document.Type.PLAIN_TEXT).build();
+
+    LanguageServiceClient languageService = LanguageServiceClient.create();
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: It's better to call it "client" instead of "service" :)

**Review by ChangchengXiao — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done, thanks!

**Pull request merged**